tools/inc/stljobs.h

```diff
@@ -0,0 +1,606 @@
+        switch (ioResult) {
+        case ERROR_SUCCESS:
+        case ERROR_OPERATION_ABORTED:
+            break;
+        default:
+            api_failure("StartThreadpoolIo callback", ioResult);
+            break;
+        }
+
+        const auto this_ = static_cast<output_collecting_pipe*>(thisRaw);
+        this_->validTill += bytes;
+        if (this_->running) {
+            this_->read_some();
+        }
+    }
+
+    void resize_create_target_buffer() {
+        if (bufferSize < targetBuffer.size() - validTill) {
+            // already has enough space
+            return;
+        }
+
+        targetBuffer.resize(validTill + bufferSize);
+        targetBuffer.resize(std::min(static_cast<size_t>(ULONG_MAX), targetBuffer.capacity()));
+    }
+
+    void read_some() {
+        assert(running.load());
+        DWORD bytesRead = 0;
+        readIo.start_threadpool_io();
+        while (resize_create_target_buffer(),
+            ReadFile(readIo.get_file(), targetBuffer.data() + validTill,
+                static_cast<unsigned long>(targetBuffer.size() - validTill), &bytesRead, &overlapped)) {
+            validTill += bytesRead;
+        }
+
+        const auto lastError = GetLastError();
+        if (lastError == ERROR_IO_PENDING) {
+            return;
+        }
+
+        readIo.cancel_threadpool_io();
+        api_failure("ReadFile", lastError);
+    }
+
+    std::atomic<bool> running{};
+    std::string targetBuffer; // if running, owned by a threadpool thread, otherwise owned by the calling thread
+    std::size_t validTill{};
+    handle<invalid_handle_value_policy> writeHandle;
+    tp_io readIo;
+    OVERLAPPED overlapped{};
+};
+
+struct execution_result {
+    unsigned long exitCode;
+    std::string output;
+};
+
+class environment_block {
+public:
+    environment_block() {
+        default_block defaultBlock;
+        auto blockEnd = static_cast<const wchar_t*>(defaultBlock.blockRaw);
+        for (;;) {
+            const auto thisLen = wcslen(blockEnd);
+            blockEnd += thisLen;
+            ++blockEnd;
+
+            if (thisLen == 0) {
+                break;
+            }
+        }
+
+        block.assign(static_cast<const wchar_t*>(defaultBlock.blockRaw), blockEnd);
+    }
+
+    [[nodiscard]] void* get() noexcept {
+        return block.data();
+    }
+
+    void append_environment(const std::wstring_view key, const std::wstring_view value) {
+        block.reserve(block.size() + key.size() + value.size() + 2);
+        block.append(key);
+        block.push_back('=');
+        block.append(value);
+        block.push_back('\0');
+    }
+
+private:
+    struct default_block {
+        void* blockRaw;
+        default_block() {
+            if (!CreateEnvironmentBlock(&blockRaw, NULL, FALSE)) {
+                api_failure("CreateEnvironmentBlock");
+            }
+        }
+
+        ~default_block() noexcept {
+            DestroyEnvironmentBlock(blockRaw);

```

Comment:
AFAIK it can fail in the same way CloseHandle can fail; i.e. only with totally bogus input. I think we should be consistent with CloseHandle handling and not address unrecoverable "the API is broken" type failures.

Reply:
It is not so much "the API is broken", but the process state is corrupt.
Checking errors in CloseFile or DestroyEnvironmentBlock (logging and exiting) may help to discover latent bugs that happened earlier that corrupted the process.